### ISSUE TITLE
AddressSanitizer complains when stopping fuzzer w/ CTRL+C

### ISSUE THREAD

**Issue by phlip9** [loiclec/fuzzcheck-rs#30]
When canceling a running fuzzer with CTRL+C, I get a big ASAN dump regardless of the fuzzing target. The repro command below uses the `usage_tests` crate, but any test function will do, even an empty one. I haven't spent any time debugging whether this is a real issue or just some ASAN false positive. Unfortunately I'm not getting symbols from AddressSanitizer on where the issue might be : (

```bash
~/dev/fuzzcheck/usage_tests/basic_crate$ RUST_BACKTRACE=1 cargo +nightly-2022-07-21 fuzzcheck --address-sanitizer tests::fuzz

    Finished release [optimized] target(s) in 12.65s
     Running unittests src/lib.rs (target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335)

running 1 test
test tests::fuzz ... 0ms  START
29ms  1 simplest_cov(0 cov: 0/46 cplx: 0.00) diverse_cov_20(0) max_each_cov_hits(0 sum: 0) diverse_cov_1(0) max_total_cov_hits(0) iter/s 38461
30ms  2 simplest_cov(1 cov: 6/46 cplx: 25.00) diverse_cov_20(6) max_each_cov_hits(1 sum: 16) diverse_cov_1(6) max_total_cov_hits(16) iter/s 4192
30ms  3 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 3911
34ms  FINISHED READING CORPUS
35ms  PULSE 292 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 570312
36ms  PULSE 584 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 422270
37ms  PULSE 1168 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 387267
41ms  PULSE 2336 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 343832
49ms  PULSE 4672 simplest_cov(2 cov: 7/46 cplx: 24.00) diverse_cov_20(7) max_each_cov_hits(2 sum: 22) diverse_cov_1(6) max_total_cov_hits(16) iter/s 326348

^C

~/dev/fuzzcheck/usage_tests/basic_crate$ =================================================================
==428240==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x625000013fa0 at pc 0x55c662d3d0c2 bp 0x625000013f80 sp 0x625000013750
WRITE of size 16 at 0x625000013fa0 thread T0
    #0 0x55c662d3d0c1  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1290c1) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #1 0x55c662ff7ab7  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x3e3ab7) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #2 0x55c662f46c98  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x332c98) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #3 0x55c662e441b6  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x2301b6) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #4 0x55c662e20f56  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x20cf56) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #5 0x55c662f1e3b4  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x30a3b4) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #6 0x55c662f0b6de  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x2f76de) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #7 0x7f4d5eee651f  (/lib/x86_64-linux-gnu/libc.so.6+0x4251f) (BuildId: 89c3cb85f9e55046776471fed05ec441581d1969)
    #8 0x55c662d8d419  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x179419) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #9 0x55c662d8d30d  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x17930d) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #10 0x55c662dd430e  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1c030e) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #11 0x55c662e4a668  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x236668) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #12 0x55c662e3e717  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x22a717) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #13 0x55c662e00498  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1ec498) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #14 0x55c662e392ea  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x2252ea) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #15 0x55c662e8e0c2  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x27a0c2) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #16 0x55c662e8cdfb  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x278dfb) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #17 0x55c662e8bfbf  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x277fbf) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #18 0x55c662e8a735  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x276735) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #19 0x55c662e84fdb  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x270fdb) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #20 0x55c662e6da8f  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x259a8f) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #21 0x55c662e82ede  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x26eede) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #22 0x55c662e83dc3  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x26fdc3) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #23 0x55c662dd9a72  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1c5a72) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #24 0x55c662dfd118  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1e9118) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #25 0x55c662fee0b9  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x3da0b9) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #26 0x55c662df8e91  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1e4e91) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
    #27 0x7f4d5eecdd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 89c3cb85f9e55046776471fed05ec441581d1969)
    #28 0x7f4d5eecde3f  (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 89c3cb85f9e55046776471fed05ec441581d1969)
    #29 0x55c662d1b144  (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x107144) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)

Address 0x625000013fa0 is a wild pointer inside of access range of size 0x000000000010.
SUMMARY: AddressSanitizer: heap-buffer-overflow (~/dev/fuzzcheck/usage_tests/basic_crate/target/fuzzcheck/x86_64-unknown-linux-gnu/release/deps/basic_crate-355b4a7b4d1de335+0x1290c1) (BuildId: faf06c4db8d903caa5edfc5b0ddaf8fc28b78ab2)
Shadow bytes around the buggy address:
  0x0c4a7fffa7a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffa7b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffa7c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffa7d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4a7fffa7e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c4a7fffa7f0: fa fa fa fa[fa]fa fa fa fa fa fa fa f1 f1 f1 f1
  0x0c4a7fffa800: f8 f2 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f2
  0x0c4a7fffa810: f2 f2 f2 f2 f8 f8 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2
  0x0c4a7fffa820: f8 f8 f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f2 f2
  0x0c4a7fffa830: f2 f2 f8 f8 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8
  0x0c4a7fffa840: f8 f8 f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==428240==ABORTING
```

possibly relevant platform details:

```bash
$ uname -a
Linux XXX 5.17.15-76051715-generic #202206141358~1655919116~22.04~1db9e34 SMP PREEMPT Wed Jun 22 19 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by loiclec** [comment]
Hi Philip ! Thanks a lot for reporting this issue. I am on holiday at the moment but I’ll look into it from next Wednesday. I will try to publish a fix as soon as I can.

**Comment by phlip9** [comment]
No rush! Enjoy your vacation : )

**Comment by loiclec** [comment]
So there were a few bugs, that I have hopefully fixed correctly in https://github.com/loiclec/fuzzcheck-rs/commit/3050ff801352d1c1e9116f6591f254ad62e07001

1. Buffer overflow: The signal handler function had the wrong signature 🤦  Only detected on macOS-x86 with AddressSanitizer on, but probably a real problem everywhere.
2. Leak: The stack that was allocated for the signal handler to run on was not deallocated after the fuzz test finished. Not a real problem for the common use case of fuzzcheck, but I fixed it anyway.
3. Buffer/Stack overflow: The stack-for-signal-handler was very small on Linux, leading to a buffer/stack overflow. This is probably the one you hit and it is also a real problem. I have increased the size of the stack to a minimum of 2^16 (~65,000) bytes. Hopefully that is enough.

Thanks again for reporting the bug. It is typically not the kind of things I would find by myself! Could you verify that it is fixed on your computer? I never got the exact same ASAN message that you did, so I am curious.

I will release a new version of fuzzcheck in a few days, when I am a bit more confident about the quality of the fix 🙂

**Comment by phlip9** [comment]
Awesome! Can confirm, looks like the fix is working (on linux x86-64)
